### PR TITLE
Integrate Cannon

### PR DESCRIPTION
**Pull request by dbeal-eth**
After success at EthDenver, we would like to add Cannon to Synthetix

Cannon has many benefits which it can offer for Synthetix immediately or near-future:
* Integration tests will become easier to maintain by isolating the chain build process to a standard definition and reducing required rebuilding of the test environment
* Any Frontend developers can easily rip up a customized, realistic testing environment. No forking, no test accounts, very little configuration
* Downstream protocols can much more easily build and extend upon our platform, and indentify problems with upcoming releases through their own tests

In addition to simply adding Cannon, there are a couple of further reaching changes which can impact:
* Synthetix has been converted to use `HardhatRuntimeEnvironment` for provider resolution, and all flags for setting network will become obsolete in favor of hardhat provider resolution. This is necessary for the building of the Cannon deployments
* L1 integration tests have been converted to build the environment with cannon rather than its own custom build. This ensures consistency with the integration tests and what we are pushing out with cannon for packages
*  (todo) CI has been updated to build cannon release and push it

cc/ @noahlitvin @mjlescano

### CODE REVIEW

File: cannonfile.toml [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+[run.synthetix]
+exec = "./publish/cannon.js"
+func = "deploy"
+modified = ["contracts", "publish/src", "publish/deployed/local"]
+
+[invoke.set_ownership]
+target = [
+    "AddressResolver",
+    "CollateralErc20",
+    #"CollateralEth",
```
Comment (barrasso) [inline cannonfile.toml:29]: how come some of these contracts are commented out?

side note: what's the purpose of this file?

File: publish/deployed/local-ovm/config.json [hunk 2]
```diff
@@ -50,10 +47,10 @@
 	"Issuer": {
 		"deploy": true
 	},
-	"Liquidator": {
+	"EternalStorageLiquidations": {
```
Comment (barrasso) [inline publish/deployed/local-ovm/config.json:50]: reverting to pre-148 liquidation mechanism?

File: publish/deployed/local/synths.json [hunk 3]
```diff
@@ -1,147 +1,74 @@
...
+		"sign": "$",
+		"description": "US Dollars",
+		"name": "sREDEEMER"
 	},
 	{
-		"name": "sADA",
-		"asset": "ADA"
+		"asset": "USD",
+		"category": "forex",
+		"sign": "$",
```
Comment (barrasso) [inline publish/deployed/local/synths.json:52]: that's a lot of redeemers bruh

File: publish/src/Deployer.js [hunk 4]
```diff
@@ -44,7 +44,11 @@ class Deployer {
 		this.nonceManager = nonceManager;
 		this.useOvm = useOvm;
 
-		this.provider = new ethers.providers.JsonRpcProvider(providerUrl);
+		if (providerUrl) {
+			this.provider = new ethers.providers.JsonRpcProvider(providerUrl);
+		} else {
+			this.provider = require('hardhat').ethers.provider;
```
Comment (barrasso) [inline publish/src/Deployer.js:50]: why is this necessary?

**Review by barrasso — COMMENTED** [review]
had some questions

File: cannonfile.toml (hunk 1 above)
Comment (dbeal-eth) [reply cannonfile.toml:29]: `cannonfile.toml` is the main file that you use in a cannon-compatible project to define your protocol (see here: https://usecannon.com/docs "Create cannonfile.toml")

the reason the lines are commented out is because theose contracts do not support changing owner (aka they don't extend `Owned`)

File: publish/deployed/local-ovm/config.json (hunk 2 above)
Comment (dbeal-eth) [reply publish/deployed/local-ovm/config.json:50]: thanks for catching, was a merge mistake

File: publish/src/Deployer.js (hunk 4 above)
Comment (dbeal-eth) [reply publish/src/Deployer.js:50]: another legacy thing, previously cannon used the hardhat network provider which was not supported as input, but now its back to using an external provider so this can be removed.

File: publish/deployed/local-ovm/.gitignore [hunk 5]
```diff
@@ -1,3 +1,4 @@
 # these files are irrelevant outside of a single developer's machine,
 # so no need to persist them in source control
-owner-actions.json
\ No newline at end of file
+deployment.json
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply publish/deployed/local-ovm/.gitignore:3]: no

File: cannonfile.toml (hunk 1 above)
Comment (barrasso) [reply cannonfile.toml:29]: > the reason the lines are commented out is because theose contracts do not support changing owner (aka they don't extend Owned)

But all `Collateral` variants are `Owned` and some (e.g. `CollateralErc20`) are not commented out

Comment (dbeal-eth) [reply cannonfile.toml:38]: let me try it again

Comment (dbeal-eth) [reply cannonfile.toml]: uncommented, let me know ify ou see any other contracts which shouldn't have been commented

**Review by barrasso — APPROVED** [review]
